kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/RecordEncryptionFilter.java

```diff
@@ -102,18 +107,18 @@ else if (isEncryptionException(throwable.getCause())) {
 
     private CompletionStage<ProduceRequestData> maybeEncodeProduce(ProduceRequestData request, FilterContext context) {
         var topicNameToData = request.topicData().stream().collect(Collectors.toMap(TopicProduceData::name, Function.identity()));
-        CompletionStage<Map<String, K>> keks = filterThreadExecutor.completingOnFilterThread(kekSelector.selectKek(topicNameToData.keySet()));
+        CompletionStage<TopicNameKekSelection<K>> keks = filterThreadExecutor.completingOnFilterThread(kekSelector.selectKek(topicNameToData.keySet()));
         return keks // figure out what keks we need
-                .thenCompose(kekMap -> {
-                    var futures = kekMap.entrySet().stream().flatMap(e -> {
+                .thenCompose(kekSelection -> {
+                    Set<String> unresolvedTopicNames = kekSelection.unresolvedTopicNames();
+                    if (!unresolvedTopicNames.isEmpty() && unresolvedKeyPolicy == UnresolvedKeyPolicy.REJECT) {
+                        return CompletableFuture.failedFuture(new UnresolvedKeyException("failed to resolve key for: " + unresolvedTopicNames));
+                    }
+                    var futures = kekSelection.topicNameToKekId().entrySet().stream().flatMap(e -> {
                         String topicName = e.getKey();
                         var kekId = e.getValue();
                         TopicProduceData tpd = topicNameToData.get(topicName);
                         return tpd.partitionData().stream().map(ppd -> {
-                            // handle case where this topic is to be left unencrypted
-                            if (kekId == null) {
```

Comment:
I introduced the TopicNameKekSelection class so that we wouldn't awkwardly have this code block specific to the passthrough behaviour that you could enter even if your policy was REJECT. Instead the decision is in one place, we look at the selection and can decide to proceed with the encryption or not.

Reply:
That seams reasonable but shouldn't `TopicNameKekSelection` also embed the policy decision? We still split the logic between two classes.

Reply:
I don't see it. The Selection is where we have Selected a Key for each input topicName (or found there is no Key for that topic).

The Policy is what we should do with that Selection, should we continue or not.

But the whole idea of making a decision this way is a stopgap, we don't want the KMS response to be involved in the decision ultimately. I wonder if we could make the implementation more future-oriented. Shaping it so that we have a legacy path, and a future path.

Maybe if the user set a property like `defaultTopicPolicy: REQUIRE_ENCRYPTION`, that would be more oriented towards where we want to go. It's not key-value based, but it's in the right direction.

Reply:
then the logic would be more like, gather up the topic names in the request, determine which should have a key (all of them), ask the selector for keys and fail out if none were resolved.